### DISCUSSION TITLE
[Optimization] Lack of directional lighting pattern / The source of 90% of shadow-casting lights

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#9810] [Ideas and suggestions]
A feature request _(maybe it should be a discussion idk)_ regarding a pretty huge issue when it comes to the amount of shadow-casting lights in submarines:

These are all the available lighting textures from lamps:
![image](https://user-images.githubusercontent.com/73229309/187901273-2c5b5649-2a61-45dd-a677-5890f82c8582.png)

Those textures are either a perfect circle, either a cone with roughly a 40° angle.
Most lamps use a circular lighting pattern, shining light in all directions, which forces builders to use shadow-casting as not to bleed light through walls.

Lamps are often placed on the ceiling with shadow casting light.
Here is what you could achieve with a 180° directional angle
![image](https://user-images.githubusercontent.com/73229309/187901121-97043c3b-56be-45af-8c1d-352d6f504ed0.png)

If you wanted to light an entire room without using shadow-casting, it'd be much more efficient:
![image](https://user-images.githubusercontent.com/73229309/187902838-99f78f28-7902-46ae-843b-92bb2c7eb452.png)

**By using directional light textures, you can significantly reduce the amount of shadow-casting lights inside subs**
How? Introducing new directional lights items, or the option to change the light texture use in Lightcomponent settings.

**Comment by MaaiG** [comment]
I support this! The problem of light sources is quite pressing (1 - https://github.com/Regalis11/Barotrauma/issues/9435, 2 - https://github.com/Regalis11/Barotrauma/discussions/9308, https://github.com/Regalis11/Barotrauma/issues/9364 ). But this proposal does not cover all the options for placement of lamps. For esample - The lamp can be placed in the corner of the room. 
It may be so: 
![187902838-99f78f28-7902-46ae-843b-92bb2c7eb452](https://user-images.githubusercontent.com/93348499/188223024-304b2dc0-3f14-4ff2-8411-160ea2b3c2ca.png)
Of course, it would be better if the [sub builder could set the light angle of the lamp by himself](https://github.com/Regalis11/Barotrauma/discussions/8052).

But in any case,  I like this idea. This is a real way to solve the problem because my enthusiasm has been greatly diminished due to all these limitations imposed by the game engine.

Finally, the section "Type of lighting" can be added to the lamp settings with a selection of parameters:
- 360° - full circle
- 180° - hulf up
- 180° - hulf down
- 180° - hulf to the right
- 180° - hulf to the left
- 90° - down and right (as in attached image)
- 90° - down and left
- 90° - up and right
- 90° - up and left

![188117654-c1e5cb87-71e6-4928-a575-518700946668](https://user-images.githubusercontent.com/93348499/188132371-17049d4e-fda6-4c41-ad0d-a221e73e37b6.png)

The main thing is that all this appeared and developed faster!

**Comment by Ek-Videogames** [comment]
Yeah I explored something like this when the update happened that made lights laggier but came to more or less the same conclusion that it was impossible to reasonably account for the number of different light dimensions needed even on a relatively tame submarine compartmentalization.

**Reply by gerakis** [reply]
The way Unity does it their lights have a built in degrees setting. They even have an inner angle and an outer angle.

 I know the game isnt made in unity but perhaps they could copy them? That way they dont have to do a collision check to cut it off.

**Comment by MaaiG** [comment]
If the ideal solution cannot be realized, it will be much better to have at least these [9 options](https://github.com/Regalis11/Barotrauma/discussions/9810#discussioncomment-3533511) (for lamps). Given that most of them are inverted or mirrored, there are only three forms.
![188117654-c1e5cb87-71e6-4928-a575-518700946668](https://user-images.githubusercontent.com/93348499/188131089-c1315202-2d77-48ae-a886-debfa8304942.png)
This will greatly facilitate the life of the sub-builder... This will really help me.
I agree that it would really allow you to get rid of 90% of the shadow-casting lights! And that's great!

**Comment by Ek-Videogames** [comment]
Here's an idea: 
- What if there's a setting for lights that clamps their bounds to the hull volume, therefore accomplishing pretty much the same effect in single-hull rooms as you would with shadows.
- Maybe for multi-hull rooms it takes linked volumes into consideration, however that might be slightly more resource intensive akin to just having shadows to begin with.

*(Also, can lights outside the submarine render correctly finally?)*

**Comment by BlackMoons** [comment]
Maaig's 3 forms + just a rotation for 'light emission' vs 'object' would take care of 80% of shadow casting light needs, and then shadow casting can just be used for anything that is left.

**Comment by BlackMoons** [comment]
> Yeah I explored something like this when the update happened that made lights laggier but came to more or less the same conclusion that it was impossible to reasonably account for the number of different light dimensions needed even on a relatively tame submarine compartmentalization.

Perfection is the enemy of doing anything at all. If directional lights can replace even half of the shadow casting lights, ie the mundane 'lets light a boring rectangular room, or the outside hull of a ship', that is a 50% improvement in shadow casting light performance right there, for little more then adding some new light textures and a dropdown box to select them.

**Comment by BlackMoons** [comment]
So instead of having dozens of different light images... How about a procedural generated light?

If lighting textures where created inside of a pixel shader on the fly, every light could be rendered in a single draw call, with no texture changes needed. Put the lighting parameters into the vertex format, since each light is only 4 vertices its not a huge overhead to stuff extra data into it.

Then everyone could directly set the start/end angle of the light cone as needed in the sub editor (like turret start/end cones), and even set 4 dimensions for the light to fade out at (Like how motion sensors have dimensions), without there needing to be a light texture for every possibility that people want. It could even save pixel fill by rendering a quad that only covers the dimensions instead of every light just being a huge square.

Throw in a parameter that selects between a few falloff types and you'd have such amazing levels of lighting customization and better lighting performance.

**Comment by girlyguppy** [comment]
they still need to look at this

**Reply by FriedrichNietzsche8** [reply]
they truly do
